jax/interpreters/pxla.py

```diff
@@ -425,10 +425,13 @@ class ShardedDeviceArray(ShardedDeviceValue, xla.DeviceArray):
   represent distinct logical shards. The correspondence can be computed with
   the assign_shards_to_replicas function.
   """
-  __slots__ = ["device_buffers", "axis_size"]
+  __slots__ = ["device_buffers", "axis_size", "aval"]
   _collect = staticmethod(onp.stack)
 
   def __init__(self, aval, device_buffers):
+    # aval must be a ShapedArray instance, because the aval_mapping rules
```

Comment:
Want to assert this rather than having a comment? Could have `assert core.skip_checks or isinstance(aval, ShapedArray)` if you want to skip the `isinstance` check by default. (In OSS we can actually prune out assert statements completely with the `-O` flag to Python.)

Reply:
I considered this, but I was reluctant to add code to the critical path of building a `ShardedDeviceArray`, which is something that happens 1000 times per step in this benchmark. WDYT?

(In fact, I have another PR coming that deletes more code from here.)

Reply:
SGTM